### PR TITLE
Rivera Leanne 

### PR DESCRIPTION
**Pull request by leannerivera**
# API Muncher
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
How did you go about exploring the Edamam API, and how did you try querying the API? | using postman and printing a lot of the data. also read docs
What is an API Wrapper? Why is it in `lib`? How would your project change if you needed to interact with more than one API (aka more than just the Edamam API)? | it is a 'faux-model' that we create to reduce the impact of any API changes to our code. I think if more than one API was needed, you can just create a new lib file and call it like the first.
Describe your API wrapper, the methods you created in it, and why you decided to create those methods/how they were helpful |  the wrapper gets all the info from the api and then parses the data. it also requests data from the api using a url formula.
What was an edge case or failure case test you wrote for your API Wrapper? What was a nominal case? | empty search queries, invalid single recipe search query are edge cases. returning back results of search query is a nominal.
How does VCR aid in testing an API? | 'records' information from the API the first go-around so that the API 'cost' is not increased while testing.
What is the Heroku URL of your deployed application? | https://chomp-crunch-munch.herokuapp.com/

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,8 @@
+Rails.application.routes.draw do
+  get root "search#new"
+  get "recipes/", to: "recipes#index", as: "recipes_index"
+  get 'recipes/:uri/view', to: "recipes#show", as: "recipe_show"
+
+  get '/search', to: 'search#new', as: "search_recipes" #where searches are routed to
```
Comment (dHelmgren) [inline config/routes.rb:6]: I'm not sure that `new` is a good name for this route/action. You're not really creating a new recipe, you're doing a search.

File: config/routes.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+Rails.application.routes.draw do
+  get root "search#new"
+  get "recipes/", to: "recipes#index", as: "recipes_index"
+  get 'recipes/:uri/view', to: "recipes#show", as: "recipe_show"
```
Comment (dHelmgren) [inline config/routes.rb:4]: Could you use `resources` for the index and show pages?

File: lib/edamam_api_wrapper.rb [hunk 3]
```diff
@@ -0,0 +1,85 @@
+require 'httparty'
+
+class EdamamApiWrapper
+   #example url search: "https://api.edamam.com/search?q=chicken&app_id=${YOUR_APP_ID}&app_key=${YOUR_APP_KEY}&from=0&to=3&calories=591-722&health=alcohol-free"
+
+  BASE_URL = "https://api.edamam.com/search" #what query params are attached to (?q for eng, ?r for nums)
+  REQUEST_URI = URI::encode("http://www.edamam.com/ontologies/edamam.owl#recipe_")
```
Comment (dHelmgren) [inline lib/edamam_api_wrapper.rb:7]: Good work distinguishing these elements from one another.

File: lib/edamam_api_wrapper.rb [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    recipe_id,
+    api_params["recipe"]["image"],
+    api_params["recipe"]["ingredientLines"],
+    api_params["recipe"]["yield"],
+    api_params["recipe"]["source"],
+    api_params["recipe"]["dietLabels"],
+    api_params["recipe"]["healthLabels"],
+    api_params["recipe"]["calories"],
+    api_params["recipe"]["totalTime"],
+    api_params["recipe"]["url"],
```
Comment (dHelmgren) [inline lib/edamam_api_wrapper.rb:80]: Could you dry this code by removing the recipe wrapper before passing the data into `single_recipe`?

File: app/controllers/recipes_controller.rb [hunk 5]
```diff
@@ -0,0 +1,23 @@
+class RecipesController < ApplicationController
+  before_action :find_query
+
+  def index
+    @recipes = EdamamApiWrapper.search_recipes(find_query).paginate(page: params[:page], per_page: 10)
```
Comment (dHelmgren) [inline app/controllers/recipes_controller.rb:5]: You've written your API wrapper to raise an error if the API call fails, but you're not looking for an error here. You should wrap this call in a `begin`/`rescue`.

This causes your pages to break on bogus data, rather than redirect.

File: app/controllers/recipes_controller.rb [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+      render :notfound, status: :not_found
+    end
+
+  end
+
+  def show
+    @recipe = EdamamApiWrapper.recipe_contents(params[:uri])
+
+    if @recipe.nil? || @recipe == false
+      redirect_to root_path, status: :not_found
```
Comment (dHelmgren) [inline app/controllers/recipes_controller.rb:17]: Your redirect doesn't complete, possibly because you're explicitly setting the status code here.

File: lib/edamam_api_wrapper.rb [hunk 7]
```diff
@@ -0,0 +1,85 @@
+require 'httparty'
+
+class EdamamApiWrapper
+   #example url search: "https://api.edamam.com/search?q=chicken&app_id=${YOUR_APP_ID}&app_key=${YOUR_APP_KEY}&from=0&to=3&calories=591-722&health=alcohol-free"
```
Comment (dHelmgren) [inline lib/edamam_api_wrapper.rb:4]: Nice work not giving away your app key, and keeping handy data nearby. Others goofed on this.

File: test/cassettes/single_recipe.yml [hunk 8]
```diff
@@ -0,0 +1,173 @@
+---
+http_interactions:
+- request:
+    method: get
+    uri: https://api.edamam.com/search?app_id=73a32603&app_key=fc79540af556f65253707c3c2b00862e&r=http://www.edamam.com/ontologies/edamam.owl%23recipe_a53ef6c8495adcb9f2859b1e5d99e9ba
```
Comment (dHelmgren) [inline test/cassettes/single_recipe.yml:5]: You didn't hide your app key from the cassettes here. :'(

File: test/controllers/recipes_controller_test.rb [hunk 9]
```diff
@@ -0,0 +1,33 @@
+require "test_helper"
+
+describe RecipesController do
+  let (:uri) {"a53ef6c8495adcb9f2859b1e5d99e9ba"}
+
+  describe "index" do
+
+    it "should get index" do
```
Comment (dHelmgren) [inline test/controllers/recipes_controller_test.rb:8]: There are some interesting cases you're not covering here:

- What happens if no search term is provided?
- What if the search returns no results?

It might also be worthwhile to add some tests around the paging parameters:

- What happens if they're absent?
- Do you get different results when they change?
- What if you send a bogus value, like a negative page number?

**Comment by dHelmgren** [comment]
# API Muncher
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Core Requirements** |
Git hygiene | Good
Comprehension questions | good
**General** |
Rails fundamentals (RESTful routing, use of named paths) | Yesish, see comments
Semantic HTML | Good
Errors are reported to the user | Could be better, see comments
API Wrapper to handle the API requests | yes
Controller testing | could be more robust, see comments
Lib testing | TOP NOTCH
Search Functionality | yes
List Functionality | yes
Show individual item functionality | yes
**Styling** |
List view shows 10 items at a time and/or has pagination | Good
Attractive and usable UI | This is in the running for my favorite style on this project! Nice work!
**API Features** |
The App attributes Edamam | Yes, except for recipe detail page.
The VCR cassettes do not contain the API key | No.
**External Resources** |
Link to deployed app on Heroku | Yes
**Overall** | This site is really beautiful! You hit most of the learning goals, but the lack of good error handling is really what got you. Don't forget to try and break your site live!
